lib/eventFactory.js

```diff
@@ -2,9 +2,98 @@
+ * @method createBuilds
+ * @param  {Pipeline} pipeline                   Pipeline to create builds
+ * @param  {Object}   config
+ * @param  {String}   config.sha                 SHA this project was built on
+ * @param  {String}   config.username            Username of the user that creates this event
+ * @param  {String}   config.scmContext          The scm context to which user belongs
+ * @param  {String}   [config.prRef]             Ref if it's a PR event
+ * @param  {String}   [config.prNum]             PR Number if it's a PR event
+ * @param  {String}   [config.startFrom]         Where the event starts from (jobname or ~commit, ~pr, etc)
+ *                                               Making optional for backward compatibility
+ * @param  {Number}   eventId                    Event's id
+ */
+function createBuilds(pipeline, config, eventId) {
+    const startFrom = config.startFrom;
+
+    // If no startFrom, then do nothing. Remove once we switch to new workflow design
+    if (!startFrom) {
+        return null;
+    }
+
+    /* eslint-disable global-require */
+    const BuildFactory = require('./buildFactory');
+    const buildFactory = BuildFactory.getInstance();
+    const JobFactory = require('./jobFactory');
+    const jobFactory = JobFactory.getInstance();
+    /* eslint-enable global-require */
+
+    const createJob = (job, jobName) => jobFactory.create({
+        pipelineId: job.pipelineId,
+        name: jobName,
+        permutations: job.permutations
+    });
+
+    return pipeline.jobs.then((jobs) => {
+        const jobNamesArray = jobs.map(j => j.name);
+        const jobsToCreate = [];
+        let jobsToStart = [];
+
+        jobs.forEach((job) => {
+            const isEnabled = job.state === 'ENABLED';
+            const isJobName = !joi.validate(startFrom, jobConfigSchema.jobname).error;
+            const isTrigger = !joi.validate(startFrom, jobConfigSchema.trigger).error;
+            const jobName = job.name;
+            const requires = job.permutations.requires;
+
+            // Do nothing if job is disabled
+            if (isEnabled) {
+                // If startFrom is a jobname, then start that job
+                if ((isJobName && startFrom === jobName)) {
+                    jobsToStart.push(job);
+                // If startFrom is a trigger (example: ~commit, ~pr), then start jobs that include that key
+                } else if (isTrigger && requires && requires.includes(startFrom)) {
+                    const isPRJob = (jobName).startsWith('PR-');
+                    const prJobName = `PR-${config.prNum}-${jobName}`;
+                    const prJobIndex = jobNamesArray.indexOf(prJobName);
+
+                    // If startFrom is not ~pr, or if ~pr and PR job already exists,
+                    // or if ~pr and isPRJob already, start job
+                    if (startFrom !== '~pr' || prJobIndex >= 0 || isPRJob) {
```

Comment:
Referring to  previous explanation (See last scenario https://github.com/screwdriver-cd/models/pull/195#discussion_r143594530)  and then this implementation, I think it will create one extra job if it's re-run.

Let's say we have a job with name `JOB_REQUIRES_PR`. If my understanding is correct then, during first run `createJob` creates a new job with name  `PR-1-JOB_REQUIRES_PR`. Now for what ever reason if this event is processed again, when job with name `JOB_REQUIRES_PR` is processed it will result in pushing  `PR-1-JOB_REQUIRES_PR` to `jobsToStart` and so does job with name `PR-1-JOB_REQUIRES_PR`, which results in 2 entries of `PR-1-JOB_REQUIRES_PR` in `jobsToStart` Array.

So I think modifying it like this will do it.

```
          // If true then it's some case of retry scenario, we should ignore it since whichever job in jobs created it, will process it again.
          const isPrJob = jobName.startsWith('PR-'); 

          if (isEnabled && !prJob) {
                // If startFrom is a jobname, then start that job
                if ((isJobName && startFrom === jobName)) {
                     jobsToStart.push(job);
                // If startFrom is a trigger (example: ~commit, ~pr), then start jobs that include that key
                } else if (isTrigger && requires && requires.includes(startFrom)) {
                    const prJobName = `PR-${config.prNum}-${jobName}`;

                    // If PR job doesn't exist yet then create the job
                    if (startFrom === '~pr' && !jobNamesArray.includes(prJobIndex)) {
                        jobsToCreate.push(createJob(job, prJobName));

                    } else {
                        jobsToStart.push(job);
                    }
}
```

Reply:
This logic will lead to starting the PR job with the wrong `j.id`. After talking to Jithin, we moved the skipping logic to inside the if statement for `isTrigger`. When the PR job already exists, we'll skip. Otherwise, if the job does not have a `startFrom` that is equal to `~pr` or if it is a PR job, start the job.